crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -491,50 +491,50 @@ impl DecisionTree {
     ) -> Vec<NodeId> {
         // 1. find potential matches between the two blocks
         let mut candidates = Vec::new();
-        let l_iter = left.iter().enumerate().filter(|&i| {
-            let ins = ctx.get_instruction(*i.1);
+        let keep_call_and_store = |node_id: NodeId| -> bool {
+            let ins = ctx.get_instruction(node_id);
             matches!(ins.operation.opcode(), Opcode::Call(_) | Opcode::Store(_))
-        });
-        let mut r_iter = right.iter().enumerate().filter(|&i| {
-            let ins = ctx.get_instruction(*i.1);
-            matches!(ins.operation.opcode(), Opcode::Call(_) | Opcode::Store(_))
-        });
-        for i in l_iter {
-            let ins = ctx.get_instruction(*i.1);
-            for j in r_iter.by_ref() {
-                //while let Some(j) = r_iter.next() {
-                let jins = ctx.get_instruction(*j.1);
-                match &jins.operation {
-                    Operation::Call { func_id, returned_arrays, .. } => {
-                        // we do not support returned arrays for now
-                        if returned_arrays.is_empty() {
-                            if let Operation::Call { func_id: f2, returned_arrays: r2, .. } =
-                                &ins.operation
-                            {
-                                if r2.is_empty() && f2 == func_id {
-                                    candidates.push(Segment::new(i, j));
-                                }
-                            }
+        };
+        let l_iter = left.iter().enumerate().filter(|&i| keep_call_and_store(*i.1));
+        let mut r_iter = right.iter().enumerate().filter(|&i| keep_call_and_store(*i.1));
```

Comment:
Can the repeated .iter().enumerate().filter also go in the lambda?

Reply:
I could do this:
```
        let keep_call_and_store = |nodes: &[NodeId]| ->Vec<(usize, &NodeId)>  {
            left.iter().enumerate().filter(|&i| {
                let ins = ctx.get_instruction(*i.1);
                matches!(ins.operation.opcode(), Opcode::Call(_) | Opcode::Store(_))
            }).collect()  
        };
```
But it seems to me that it is creating a new vector, and I was hoping to just iterate over the relevant nodes. Do you have a better suggestion than mine?

Reply:
I am open to suggestions here as I could not make it without copying of the nodes.

Reply:
```rs
    fn synchronise<'a>(
        &self,
        ctx: &mut SsaContext,
        left: &'a [NodeId],
        right: &'a [NodeId],
        block_id: BlockId,
    ) -> Vec<NodeId> {
        // 1. find potential matches between the two blocks
        let mut candidates = Vec::new();
        let calls_and_stores = |ids: &'a [NodeId]| {
            ids.iter().enumerate().filter(|&(_, &id)| {
                let ins = ctx.get_instruction(id);
                matches!(ins.operation.opcode(), Opcode::Call(_) | Opcode::Store(_))
            })
        };
        let mut r_iter = calls_and_stores(right);
        for left_node in calls_and_stores(left) {
            ...
```
Seems like rust can't infer the correct `'a`

Reply:
I don't like lifetimes so I prefer the version without them.